Site Editor: Add "Added by" description to template part navigation sidebar

kevin940726 opened the pull request:
<!-- Thanks for contributing to Gutenberg! Please follow the Gutenberg Contributing Guidelines:
https://github.com/WordPress/gutenberg/blob/trunk/CONTRIBUTING.md -->

## What?
<!-- In a few words, what is the PR actually doing? -->
Close https://github.com/WordPress/gutenberg/issues/48483. Add "Added by" description to the template part navigation sidebar.

## Why?
<!-- Why is this PR necessary? What problem is it solving? Reference any existing previous issue(s) or PR(s), but please add a short summary here, too -->
See https://github.com/WordPress/gutenberg/issues/48483.

## How?
<!-- How is your PR addressing the issue at hand? What are the implementation details? -->
In order to reuse the logic in `<AddedBy>`, I extract them into a hook `useAddedBy`, and use that in the sidebar. Open to ideas to make it more simple and more elegant!

## Testing Instructions
<!-- Please include step by step instructions on how to test this PR. -->
<!-- 1. Open a Post or Page. -->
<!-- 2. Insert a Heading Block. -->
<!-- 3. etc. -->
1. Go to Site Editor
2. Click "Template parts" in the sidebar.
3. Click on any template part, and expect to see the "Added by" description.

### Testing Instructions for Keyboard
<!-- How can you test the changes by using the keyboard only? Please note, this is required for PRs that change the user interface (UI). This ensures the PR can be tested for any possible accessibility regressions. -->
1. Go to the Site Editor with template parts: `/wp-admin/site-editor.php?canvas=view&postType=wp_template_part&path=%2Fwp_template_part`
2. Press <kbd>Tab</kbd> four times to navigate to the template parts list
3. Enter either of the template part, expect the screen readers to announce the "Added by" description in the sidebar.

## Screenshots or screencast <!-- if applicable -->
![image](https://user-images.githubusercontent.com/7753001/222747958-4abd8ccd-671b-4faf-8071-23dae78012c4.png)

![image](https://user-images.githubusercontent.com/7753001/222748011-f658796c-52be-4af9-a796-cc7a029acc4d.png)

kevin940726 committed:
Add Added by description to template navigation sidebar

kevin940726 commented:
Question: Do we also want to display the icon as in the template parts list view? It could help distinguish different types of authors: theme, plugin, site, or user.

jameskoster commented:
I know we won't get there for 6.2, but it would be sweet if you could reset customisations in this panel. 

<img width="434" alt="Screenshot 2023-03-03 at 15 22 13" src="https://user-images.githubusercontent.com/846565/222758684-1b835654-4209-4506-9cee-1fc1c0359c74.png">

On that note, should we still mark customised template parts somehow? The management view does this:

<img width="201" alt="Screenshot 2023-03-03 at 15 22 49" src="https://user-images.githubusercontent.com/846565/222758835-81d83c50-b1ae-4190-a77b-9803bf5e80ab.png">

Maybe we can add a similar suffix here? "Added by Twenty Twenty-Three (customized)".

> Question: Do we also want to display the icon as in the template parts list view?

I don't think it's essential for 6.2. We need to do a little design exploration on how best to organise things there. Icons might be the solution, but we can also try splitting the list into sections, or even having dedicated panels for each semantic area.

kevin940726 commented:
> should we still mark customised template parts somehow?

I can take it for a spin!

> I don't think it's essential for 6.2. We need to do a little design exploration on how best to organise things there. 

Sure thing! Sounds good to me.

kevin940726 committed:
Add customized string

kevin940726 commented:
> should we still mark customised template parts somehow

Done in [7680c6c](https://github.com/WordPress/gutenberg/pull/48732/commits/7680c6c13a6875679d003a7397fe3f3a3b6468c1).

![image](https://user-images.githubusercontent.com/7753001/223028259-2bdfe95d-a0bf-4182-802f-0c8e8f1be54d.png)

jameskoster reviewed (approved):
This is working as expected for me. 

> Question: Do we also want to display the icon as in the template parts list view?

Going back to this point, it might be worth adding some descriptive text based on the semantic areas. But I think we can do so in a separate issue.

kevin940726 committed:
Rename to AddedByText

talldan commented on packages/edit-site/src/components/list/added-by.js:
```diff
@@ -56,128 +144,16 @@ function BaseAddedBy( { text, icon, imageUrl, isCustomized, templateType } ) {
+	const addedBy = useAddedBy( template );
 
 	return (
 		<BaseAddedBy
```
Is it worth keeping `AddedBy` and `BaseAddedBy` as separate components, or could they be merged into one?

kevin940726 replied:
Oh yeah, makes sense! Will merge them into one.

talldan commented on packages/edit-site/src/components/list/added-by.js:
```diff
@@ -20,6 +20,94 @@ import { _x } from '@wordpress/i18n';
+ * @param    {Object}      template     The template object.
+ * @return {AddedByData} The added by object or null.
+ */
+export function useAddedBy( template ) {
```
This is probably how I should have done it from the start 😄 

Thanks for improving it 🎉

youknowriad replied:
It feels a bit heavy to pass the "template" object as a prop and argument to these hooks. Maybe it's better to receive (postType, postId) couple instead like we do in similar kind of hooks.

youknowriad replied:
(passing objects is often a performance problem, as you can't optimize and have to re-render for any change in the object for the whole tree)

kevin940726 replied:
Well, it can still be optimized by wrapping the hook inside an intermediate component, like how react-redux used to do:

```js
function AddedByText = ( { postType, postId } ) {
	const { record } = useEditedEntityRecord( postType, postId );
	const addedBy = useAddedBy( record );
	return addedBy.text;
}
```

A benefit of using the template object is that we don't have to depend on the `edit-site` store in this hook. It can be a _dumb_ hook (well, not really, as it still depends on the core data store), and a bit easier to unit test if needed.

I don't have a strong preference for this though, as I think either is fine and the performance impact doesn't matter much (until proven otherwise). Currently, the places that use this hook/component already have the `template` object in place, so I don't think it would matter in this situation.

I'd be happy to change this to `postType` and `postId` if you feel strongly about this though. I don't know much about other existing hooks, so consistency might be more important here too!

youknowriad replied:
Yeah, I'd appreciate if we change this personally 👍 

I think the performance optimization are not negligible. I remember we had a similar issue in the past where the whole editor was re-rendering because we were passing the "template" around and any change to any block changes the template (blocks), so it rerenders all the components.

youknowriad replied:
Thanks for the update here.

talldan commented on packages/edit-site/src/components/list/added-by.js:
```diff
@@ -20,6 +20,94 @@ import { _x } from '@wordpress/i18n';
+				}
+
+				// Added by plugin.
+				// Template originally provided by a plugin, but customized by a user.
```
I think the part about 'customized by a user' is incorrect (as it was in `trunk`), as we can see there's an `isCustomized` property lower down to determine this. It might be a good opportunity to remove it.

```suggestion
```

talldan reviewed (approved):
It's nice to see this tidied up, good work!

kevin940726 commented:
Seems like this has some overlaps with #48739. We have a few options:

1. Only show the description in #48739
![image](https://user-images.githubusercontent.com/7753001/223336709-35179a84-bd26-4125-b5ad-6c3c102c3491.png)

2. Only show the added by text for template parts
![image](https://user-images.githubusercontent.com/7753001/223336938-3e9b306f-d736-4814-bce1-dc17c1cf1187.png)

3. Show both
![image](https://user-images.githubusercontent.com/7753001/223337136-6307383a-d91f-4562-9346-bb340c9a6a02.png)

Which do we want to move forward to?

talldan commented:
Both could be visible, but it might need a bit of design magic. ✨

kevin940726 committed:
Code review

kevin940726 committed:
Keep both descriptions

kevin940726 commented:
I'll go with both for now. Welcome to do any design tweaks on this!

Mamaduka commented:
Thank you, @kevin940726!

I'll try to cherry-pick this for Beta 5 based on design feedback.

jameskoster commented:
It's probably fine to have both, though "This is your Footer template part" is not super useful. Maybe we concatenate? 

> This is your Footer template part, added by Twenty Twenty-Three.

For the description I'd rather push towards something like https://github.com/WordPress/gutenberg/issues/48777, but suspect we don't have time.

kevin940726 committed:
Concat descriptions

kevin940726 committed:
Refactor to postType and postId

kevin940726 committed:
Modify into a local hook

jameskoster commented:
After a quick chat with Matias:

* There's not so much value in listing the _theme_ author since that is kind of the default
* "Added by" wording isn't great and probably not necessary
* For plugins it would be great if we can use a brand element.

A revised design:

<img width="358" alt="Screenshot 2023-03-07 at 16 51 46" src="https://user-images.githubusercontent.com/846565/223492060-cb8e56a2-221d-47e5-afc8-4ab96dfa847c.png">

The icon comes from the block category supplied by the plugin. We can fallback to the generic `plugin` icon when necessary. For custom templates we can display the username + Gravatar.

kevin940726 committed:
Design tweaks

kevin940726 commented:
Here's what it looks like for theme-added customized template parts.

![image](https://user-images.githubusercontent.com/7753001/223630934-2f416433-4ec2-411a-bcc4-0f8c6a54ebb1.png)

And for user:

![image](https://user-images.githubusercontent.com/7753001/223631221-17cffad3-238e-445f-9b0f-1377ddcfd9b9.png)

kevin940726 commented:
Right now it only shows a generic plugin icon. It seems like relying on the block categories' icon is not consistent across plugins, as they are not related to the templates. Any ideas for this? Or we can enhance it in a follow-up.

![image](https://user-images.githubusercontent.com/7753001/223671689-36c92315-922f-43a6-804f-ef66856df70e.png)

jameskoster committed:
Style adjustments

- increase space between icon / label
- apply rounded corners to gravatar
- colorise fallback icon

jameskoster commented:
Thanks for the revisions.

I don't think we need to display this detail for theme-added template parts, since that's kind of the default. It seems more important to indicate when templates are _not_ from the theme, IE plugin or user created.

> It seems like relying on the block categories' icon is not consistent across plugins, as they are not related to the templates

How do you mean? I don't think it matters if they're related to templates, it's more of a branding detail that can be used in multiple places. Similar to the Plugin panel in the editor:

<img width="182" alt="Screenshot 2023-03-08 at 11 01 19" src="https://user-images.githubusercontent.com/846565/223696241-f19e208d-133a-4f6d-8bf1-b52e36cbbe01.png">

The plugin icon is fine as a fallback.

---

I noticed a small bug, if I customise a plugin template then it switches from plugin-created to user-created:

<img width="1022" alt="Screenshot 2023-03-08 at 11 08 22" src="https://user-images.githubusercontent.com/846565/223697872-082832b9-53f8-43f0-b4ce-f50beb32b07e.png">

It should still be marked as a plugin template, but also marked as customized.

kevin940726 commented:
> How do you mean? I don't think it matters if they're related to templates, it's more of a branding detail that can be used in multiple places.

The only info we get from `getCategories()` right now is `title`, `slug`, and `icon`. Categories themselves don't seem to be related to plugins (there's no `plugin` field). Moreover, a plugin can register multiple block categories, should we just choose a random one in this case? Maybe there's an API for that I don't know of though!

> I noticed a small bug, if I customise a plugin template then it switches from plugin-created to user-created

It should already be how it behaves in trunk in the manage template parts view though. Might be worth opening a new issue for that?

talldan commented:
The icon that a plugin adds for a block category might not be anything like the plugin's logo or branding. It could be a really plain 'category' icon, and it would be look like a bug to use that to represent the plugin.

The block category icon was never pitched as a way for a plugin to add its branding, it just happens to be what WooCommerce does in that example.

Mamaduka commented:
@kevin940726, @jameskoster, may I suggest punting this for WP 6.3? Sorry, I forgot that beta 5 already had a soft string freeze.

kevin940726 commented:
Fine with me 👍

jameskoster commented:
Thanks for clarifying. I guess we'll have to run with the default plugin icon to begin with, but we should explore a way for plugins to supply a branding element (probably icon + color), it will be important as we move into phase 3.

> It should already be how it behaves in trunk in the manage template parts view though. Might be worth opening a new issue for that?

That's not how it behaves on trunk. Here's what I see on the template management view:

<img width="1213" alt="Screenshot 2023-03-09 at 10 03 55" src="https://user-images.githubusercontent.com/846565/223988436-b32511a2-d827-42dd-b819-3b4f7d38a6d9.png">

kevin940726 commented:
> That's not how it behaves on trunk. Here's what I see on the template management view

Interesting! Looks like if you go to the manage templates view, then go back to single template view, it will show "_plugin_ (customized)". But as soon as editing the template, it will show the author's name instead. I'm not sure what went wrong, because it's using the exact same code as in the management template view. When will `has_theme_file` change? Anyway, I don't think it is related to this PR, but this PR does make it more visible though.

jameskoster commented:
Do you think we should fix that (separately) before merging this? The inconsistency seems a bit awkward to me.

kevin940726 commented:
> Do you think we should fix that (separately) before merging this?

Yeah, I think so, cuz I'm not sure how to fix it. Unless someone can point me in a direction of a quick fix? Maybe it's still better to separate it into a different PR so that discussions don't diverge there.

jameskoster commented:
Separate PR sounds good 👍 

Since this is no longer on the table for 6.2, I don't think it necessarily needs to be blocked.

jameskoster commented:
The last thing to do here is remove the "added by" detail when the template was supplied by the active theme.

kevin940726 committed:
Don't show added by active theme

kevin940726 commented:
@jameskoster Not sure if I understand it correctly. I just pushed a commit to hide the "added by" detail if the template is added by a theme and the theme is the currently active theme.

jameskoster reviewed (approved):
Thanks @kevin940726 ! I think this is working well. We can address the issue around plugin templates being marked as custom separately.
